--- configure.in	2014-04-16 17:04:49 -0300
+++ configure.in	2015-09-29 12:57:03 -0300
@@ -1566,7 +1566,7 @@
 dnl If it fails, test again for normal libmfhdf/libdf
   if test "$HDF_LIB_NAME" = "missing" ; then
     unset ac_cv_lib_mfhdf_SDreaddata
-    AC_CHECK_LIB(mfhdf,SDreaddata,HDF_LIB_NAME="-lmfhdf -ldf",HDF_LIB_NAME=missing,-ldf)
+    AC_CHECK_LIB(mfhdf,SDreaddata,HDF_LIB_NAME="-lmfhdf -lxdr -ldf",HDF_LIB_NAME=missing,-lxdr -ldf)
     if test "$HDF_LIB_NAME" = "missing" ; then
       unset ac_cv_lib_mfhdf_SDreaddata
       AC_CHECK_LIB(hdf4,SDreaddata,HDF_LIB_NAME=-lhdf4,HDF_LIB_NAME=missing,)
@@ -1576,13 +1576,13 @@
 dnl Test again, with -ldf, -ljpeg and -lz this time.
   if test "$HDF_LIB_NAME" = "missing" ; then
     unset ac_cv_lib_mfhdf_SDreaddata
-    AC_CHECK_LIB(mfhdf,SDreaddata,HDF_LIB_NAME="-lmfhdf -ldf",HDF_LIB_NAME=missing,-ldf -ljpeg -lz)
+    AC_CHECK_LIB(mfhdf,SDreaddata,HDF_LIB_NAME="-lmfhdf -lxdr -ldf",HDF_LIB_NAME=missing,-lxdr -ldf -ljpeg -lz)
   fi
 
 dnl Not found... again, with -lsz.
   if test "$HDF_LIB_NAME" = "missing" ; then
     unset ac_cv_lib_mfhdf_SDreaddata
-    AC_CHECK_LIB(mfhdf,SDreaddata,HDF_LIB_NAME="-lmfhdf -ldf -lsz",HDF_LIB_NAME=missing,-ldf -lsz -ljpeg -lz)
+    AC_CHECK_LIB(mfhdf,SDreaddata,HDF_LIB_NAME="-lmfhdf -lxdr -ldf -lsz",HDF_LIB_NAME=missing,-lxdr -ldf -lsz -ljpeg -lz)
   fi
 
   if test "$HDF_LIB_NAME" != "missing" ; then

--- configure	2014-04-16 17:04:49 -0300
+++ configure	2015-09-29 12:57:09 -0300
@@ -21159,7 +21159,7 @@
   $as_echo_n "(cached) " >&6
 else
   ac_check_lib_save_LIBS=$LIBS
-LIBS="-lmfhdf -ldf $LIBS"
+LIBS="-lmfhdf -lxdr -ldf $LIBS"
 cat confdefs.h - <<_ACEOF >conftest.$ac_ext
 /* end confdefs.h.  */
 
@@ -21190,7 +21190,7 @@
 { $as_echo "$as_me:${as_lineno-$LINENO}: result: $ac_cv_lib_mfhdf_SDreaddata" >&5
 $as_echo "$ac_cv_lib_mfhdf_SDreaddata" >&6; }
 if test "x$ac_cv_lib_mfhdf_SDreaddata" = xyes; then :
-  HDF_LIB_NAME="-lmfhdf -ldf"
+  HDF_LIB_NAME="-lmfhdf -lxdr -ldf"
 else
   HDF_LIB_NAME=missing
 fi
@@ -21250,7 +21250,7 @@
   $as_echo_n "(cached) " >&6
 else
   ac_check_lib_save_LIBS=$LIBS
-LIBS="-lmfhdf -ldf -ljpeg -lz $LIBS"
+LIBS="-lmfhdf -lxdr -ldf -ljpeg -lz $LIBS"
 cat confdefs.h - <<_ACEOF >conftest.$ac_ext
 /* end confdefs.h.  */
 
@@ -21281,7 +21281,7 @@
 { $as_echo "$as_me:${as_lineno-$LINENO}: result: $ac_cv_lib_mfhdf_SDreaddata" >&5
 $as_echo "$ac_cv_lib_mfhdf_SDreaddata" >&6; }
 if test "x$ac_cv_lib_mfhdf_SDreaddata" = xyes; then :
-  HDF_LIB_NAME="-lmfhdf -ldf"
+  HDF_LIB_NAME="-lmfhdf -lxdr -ldf"
 else
   HDF_LIB_NAME=missing
 fi
@@ -21296,7 +21296,7 @@
   $as_echo_n "(cached) " >&6
 else
   ac_check_lib_save_LIBS=$LIBS
-LIBS="-lmfhdf -ldf -lsz -ljpeg -lz $LIBS"
+LIBS="-lmfhdf -lxdr -ldf -lsz -ljpeg -lz $LIBS"
 cat confdefs.h - <<_ACEOF >conftest.$ac_ext
 /* end confdefs.h.  */
 
@@ -21327,7 +21327,7 @@
 { $as_echo "$as_me:${as_lineno-$LINENO}: result: $ac_cv_lib_mfhdf_SDreaddata" >&5
 $as_echo "$ac_cv_lib_mfhdf_SDreaddata" >&6; }
 if test "x$ac_cv_lib_mfhdf_SDreaddata" = xyes; then :
-  HDF_LIB_NAME="-lmfhdf -ldf -lsz"
+  HDF_LIB_NAME="-lmfhdf -lxdr -ldf -lsz"
 else
   HDF_LIB_NAME=missing
 fi
